cmd/validate-krew-manifest/main.go

```diff
@@ -0,0 +1,179 @@
+	flag.Set("logtostderr", "true") // Set glog default to stderr
+	// TODO(ahmetb) iterate over glog flags and hide them (not sure if possible without using pflag)
+	flag.Parse()
+}
+
+func main() {
+	defer glog.Flush()
+
+	if flManifest == "" {
+		glog.Fatal("-manifest must be specified")
+	}
+
+	if err := validateManifestFile(flManifest); err != nil {
+		glog.Fatalf("%v", err) // with stack trace
+	}
+}
+
+func validateManifestFile(path string) error {
+	glog.V(4).Infof("reading file %s", path)
+	p, err := indexscanner.ReadPluginFile(path)
+	if err != nil {
+		return errors.Wrap(err, "failed to read plugin file")
+	}
+	filename := filepath.Base(path)
+	pluginNameFromFileName := strings.TrimSuffix(filename, filepath.Ext(filename))
+	glog.V(4).Infof("inferred plugin name as %s", pluginNameFromFileName)
+
+	// validate plugin manifest
+	if err := p.Validate(pluginNameFromFileName); err != nil {
+		return errors.Wrap(err, "plugin validation error")
+	}
+	glog.Infof("structural validation OK")
+
+	// make sure each platform matches a supported platform
+	for i, p := range p.Spec.Platforms {
+		if os, arch := findAnyMatchingPlatform(p.Selector); os == "" || arch == "" {
+			return errors.Errorf("spec.platform[%d]'s selector (%v) doesn't match any supported platforms", i, p.Selector)
+		}
+	}
+	glog.Infof("all spec.platform[] items are used")
+
+	// validate no supported <os,arch> is matching multiple platform specs
+	if err := isOverlappingPlatformSelectors(p.Spec.Platforms); err != nil {
+		return errors.Wrap(err, "overlapping platform selectors found")
+	}
+	glog.Infof("no overlapping spec.platform[].selector")
+
+	// exercise "install" for all platforms
+	for i, p := range p.Spec.Platforms { // TODO(ahmetb) make this a testable method
+		glog.Infof("installing spec.platform[%d]", i)
+		if err := installPlatformSpec(path, p); err != nil {
+			return errors.Wrapf(err, "spec.platforms[%d] failed to install", i)
+		}
+		glog.Infof("installed  spec.platforms[%d]", i)
+	}
+	log.Printf("all %d spec.platforms installed fine", len(p.Spec.Platforms))
+	return nil
+}
+
+// isOverlappingPlatformSelectors validates if multiple platforms have selectors
+// that match to a supported <os,arch> pair.
+func isOverlappingPlatformSelectors(platforms []index.Platform) error {
+	// TODO(ahmetb) implement
+	for _, v := range allPlatforms() {
+		os, arch := v[0], v[1]
+
+		var matchIndex []int
+		for i, p := range platforms {
+			if selectorMatchesOSArch(p.Selector, os, arch) {
+				matchIndex = append(matchIndex, i)
+			}
+		}
+
+		if len(matchIndex) > 1 {
+			return errors.Errorf("multiple spec.platforms (at indexes %v) have overlapping selectors that select os=%s/arch=%s", matchIndex, os, arch)
+		}
+	}
+	return nil
+}
+
+// installPlatformSpec installs the p to a temporary location on disk to verify
+// by shelling out to external command.
+func installPlatformSpec(manifestFile string, p index.Platform) error {
+	goos, goarch := findAnyMatchingPlatform(p.Selector)
```

Comment:
@ahmetb testing this on linux. If I'm understanding correctly, this attempts to install manifest for all specified platforms. It currently fails on linux when installing for the `windows` platform: https://gist.github.com/juanvallejo/1ea80ea7406362ac9a78d517bbddb680

Reply:
Change-ns manifest is actually broken. Windows spec should have “bin” field end in .exe.